rfcs/0188-batched-label-to-task-status.md

```diff
@@ -0,0 +1,83 @@
+# RFC 188 - Adding a batched "label to task status" endpoint
+* Comments: [#188](https://github.com/taskcluster/taskcluster-rfcs/pull/188)
+* Proposed by: @Alphare
+
+# Summary
+
+Add an endpoint to query the latest status for a given task label, for each
+label in a provided list.
+
+## Motivation
+
+When looking at profiles with the following command in the `mozilla-unified`
+repo:
+
+`py-spy record -F --idle --format speedscope -o output.json -- ./mach taskgraph morphed -p taskcluster/test/params/mc-onpush.yml`
+
+I saw that the majority (near 70%) of the time was spent on the optimization
+step, as the screenshot below shows:
+
+![20231211_10h54m47s_grim](https://github.com/taskcluster/taskcluster/assets/9445758/62c400cc-a125-4f08-b7dd-c8bc9a9e9a6d)
+
+Further investigating, it turns out that basically all of this time is spent
+waiting on the network doing `read()` calls. Basically, for every task, there
+is one call to `find_task_id`, then one call to `status_task`, the goal
+of which is to get the status of the latest task that corresponds to a label.
```

Comment:
I'm trying to understand if that latest task with that label belongs to some specific task group or any group? 

There might be lots of tasks running with the same label at any given point of time?

Reply:
My understanding is that taskgraph script is trying to fetch all indexes first with given namespace, and then query those tasks returned.

If we should optimize this process, we should likely go through `indexes` service and not do that query on `tasks` table directly.  Which would be a faster and somewhat easier probably

Reply:
Yeah I think generalizing this a bit, we likely want to implement the following new APIs:
* `queue/v1/tasks`
* `queue/v1/tasks/status`

Where the tasks are submitted in the request data rather than in the path. I guess we could go all the way and expand all the APIs currently under `/task/`, but imo the above two would make the most sense and be the highest value.

I noticed that `index/api/tasks` already exists, so we can already start using this. Implementing a similar endpoint to this over on the queue service makes a lot of sense imo.

Reply:
> I noticed that index/v1/tasks already exists, so we can already start using this. Implementing a similar endpoint to this over on the queue service makes a lot of sense imo.

Nvm, it turns out `index/v1/tasks` only works for a specific namespace (not multiple namespaces). I guess we could still use it to get *every* cached task, but then we'd be going from requesting a couple thousand tasks to tens of thousands of tasks. Maybe this is still faster overall though?

Reply:
You can create multiple routes in tasks, so I think you should be able to create a custom namespace that contains exactly the tasks you are interested in, if you don't have one already? I think it should just be a case of organising a namespace to contain the index entries that are needed.

Reply:
> that contains exactly the tasks you are interested in

The tasks we are interested in changes over time. All cached tasks go into a `cache.<name>.<digest>` namespace. So we can get all cached tasks with `<name>`, or all cached tasks, but that will include all the old tasks that are either unused or used on older branches. So we'd essentially need to get the ids of all cached tasks, just to pick out a small subset of the ones we are interested in.

Like I said, maybe this is still faster given the fewer network requests. But regardless of whether this is faster or not, is there a compelling reason not to implement more efficient APIs in Taskcluster? It'll help us either way

Reply:
Plus, this still only addresses the index query. We still need this for getting the status.